apps/prairielearn/src/lib/question-submission.ts

```diff
@@ -1,32 +1,46 @@
-//@ts-check
-const _ = require('lodash');
+import { omit } from 'lodash';
+import { type Request, type Response } from 'express';
+
 import * as error from '@prairielearn/error';
+
 import { saveAndGradeSubmission, saveSubmission } from './grading';
 import { idsEqual } from './id';
 import { selectVariantById } from '../models/variant';
+import { type Variant } from './db-types';
 
-export async function validateVariantAgainstQuestion(unsafe_variant_id, question_id) {
+export async function validateVariantAgainstQuestion(
+  unsafe_variant_id: string,
+  question_id: string,
+  instance_question_id: string | null = null,
+): Promise<Variant> {
   const variant = await selectVariantById(unsafe_variant_id);
   if (variant == null || !idsEqual(variant.question_id, question_id)) {
     throw error.make(
       400,
       `Client-provided variant ID ${unsafe_variant_id} is not valid for question ID ${question_id}.`,
     );
   }
+  if (
+    instance_question_id != null &&
+    (!variant.instance_question_id || !idsEqual(variant.instance_question_id, instance_question_id))
+  ) {
+    throw error.make(
+      400,
+      `Client-provided variant id "${unsafe_variant_id}" does not belong to the authorized instance question "${instance_question_id}"`,
```

Comment:
Nit, let's make this and the above error message on line 20 consistent:

- Capitalize "id"
- Drop quotes around IDs